app/models/auth_sources/auth_source_ldap.rb

```diff
@@ -124,18 +124,22 @@ def update_usergroups(login)
     end
 
     logger.debug "Updating user groups for user #{login}"
-    internal = User.unscoped.find_by_login(login).external_usergroups.map(&:name)
-    external = ldap_con.group_list(login) # this list may return all groups in lowercase
-    (internal | external).each do |name|
-      begin
-        external_usergroup = external_usergroups.where('lower(name) = ?', name.downcase).last
-        if external_usergroup.present?
-          logger.debug "Refreshing external user group #{external_usergroup.name}"
-          external_usergroup.refresh
-        end
-      rescue => error
-        logger.warn "Could not update user group #{name}: #{error}"
-      end
+    user = User.unscoped.find_by_login(login)
+    internal = user.external_usergroups.pluck(:name).map {|g| g.downcase}
+    external = ldap_con.group_list(login).map {|g| g.downcase}
+    group_name_arel = ExternalUsergroup.arel_table[:name].lower
+
+    # Remove from internal groups already gone on auth_source side
+    user.external_usergroups.preload(:usergroup).where(group_name_arel.not_in(external)).find_each do |external_usergroup|
+      logger.debug "Removing user from external group #{external_usergroup.name}"
+      external_usergroup.usergroup.user_ids -= [user.id]
```

Comment:
I'd extract usergroup to local var to avoid repeating `external_usergroup.usergroup`, the same below